From 318c008fae84e7f85650b227e67cf72d33c8f724 Mon Sep 17 00:00:00 2001
From: Yimin Peng <yiminp@codeaurora.org>
Date: Wed, 27 Dec 2017 15:56:15 +0800
Subject: [PATCH] add support for USB pass through device on GVM

Add defconfig and driver changes to support USB
as a pass through device on GVM for MSM8996

Change-Id: Id1e9f0467bcce96ba0cca61a3844d222659fef55
Signed-off-by: Ramachandran Venkataramani <ramavenk@codeaurora.org>
Signed-off-by: Yimin Peng <yiminp@codeaurora.org>
---
 arch/arm64/configs/msm-auto-gvm_defconfig |  55 ++++++++++++
 drivers/clk/msm/clock-dummy.c             |  56 ++++++++++++-
 drivers/platform/msm/gpio-usbdetect.c     | 133 +++++++++++++++++++-----------
 drivers/usb/core/hcd.c                    |   4 +
 drivers/usb/dwc3/core.c                   |   9 +-
 drivers/usb/dwc3/dwc3-msm.c               |  10 +++
 drivers/usb/host/xhci-plat.c              |   8 ++
 drivers/usb/host/xhci.c                   |   5 ++
 drivers/usb/phy/phy-msm-ssusb-qmp.c       |  32 +++++--
 9 files changed, 251 insertions(+), 61 deletions(-)

diff --git a/arch/arm64/configs/msm-auto-gvm_defconfig b/arch/arm64/configs/msm-auto-gvm_defconfig
index c5d68e6..2aa7809 100644
--- a/arch/arm64/configs/msm-auto-gvm_defconfig
+++ b/arch/arm64/configs/msm-auto-gvm_defconfig
@@ -336,3 +336,58 @@ CONFIG_CRYPTO_MD4=y
 CONFIG_CRYPTO_ARC4=y
 CONFIG_CRYPTO_TWOFISH=y
 CONFIG_QMI_ENCDEC=y
+CONFIG_USB=y
+CONFIG_USB_XHCI_HCD=y
+CONFIG_USB_EHCI_HCD=y
+CONFIG_USB_EHCI_HCD_PLATFORM=y
+CONFIG_USB_OHCI_HCD=y
+CONFIG_USB_OHCI_HCD_PLATFORM=y
+CONFIG_USB_STORAGE=y
+CONFIG_USB_DWC3=y
+CONFIG_USB_ISP1760=y
+CONFIG_USB_QTI_KS_BRIDGE=y
+CONFIG_NOP_USB_XCEIV=y
+CONFIG_USB_MSM_SSPHY_QMP=y
+CONFIG_MSM_QUSB_PHY=y
+CONFIG_USB_ULPI=y
+CONFIG_USB_GADGET=y
+CONFIG_USB_GADGET_VBUS_DRAW=500
+CONFIG_USB_CONFIGFS=y
+CONFIG_USB_CONFIGFS_SERIAL=y
+CONFIG_USB_CONFIGFS_NCM=y
+CONFIG_USB_CONFIGFS_ECM=y
+CONFIG_USB_CONFIGFS_MASS_STORAGE=y
+CONFIG_USB_CONFIGFS_F_FS=y
+CONFIG_USB_CONFIGFS_F_MTP=y
+CONFIG_USB_CONFIGFS_F_PTP=y
+CONFIG_USB_CONFIGFS_F_ACC=y
+CONFIG_USB_CONFIGFS_UEVENT=y
+CONFIG_USB_CONFIGFS_F_DIAG=y
+CONFIG_USB_CONFIGFS_F_CDEV=y
+CONFIG_USB_CONFIGFS_F_QDSS=y
+CONFIG_GPIO_USB_DETECT=y
+CONFIG_USB_USBNET=y
+CONFIG_USB_BAM=y
+CONFIG_USB_CONFIGFS_QCRNDIS=y
+CONFIG_USB_CONFIGFS_RMNET_BAM=y
+CONFIG_USB_F_QCRNDIS=y
+CONFIG_USB_F_RMNET_BAM=y
+CONFIG_DIAG_OVER_USB=y
+CONFIG_EXTCON=y
+CONFIG_RNDIS_IPA=y
+CONFIG_SCSI=y
+CONFIG_BLK_DEV_SD=y
+CONFIG_CHR_DEV_SG=y
+CONFIG_CHR_DEV_SCH=y
+CONFIG_SCSI_CONSTANTS=y
+CONFIG_SCSI_LOGGING=y
+CONFIG_SCSI_SCAN_ASYNC=y
+CONFIG_SCSI_UFSHCD=y
+CONFIG_SCSI_UFSHCD_PLATFORM=y
+CONFIG_SCSI_UFS_QCOM=y
+CONFIG_SCSI_UFS_QCOM_ICE=y
+CONFIG_USB_DWC3_PCI=y
+CONFIG_USB_XHCI_PCI=y
+CONFIG_USB_EHCI_PCI=y
+CONFIG_USB_OHCI_HCD_PCI=y
+CONFIG_DIAG_CHAR=y
diff --git a/drivers/clk/msm/clock-dummy.c b/drivers/clk/msm/clock-dummy.c
index e874fcc..9206a6d 100644
--- a/drivers/clk/msm/clock-dummy.c
+++ b/drivers/clk/msm/clock-dummy.c
@@ -14,6 +14,51 @@
 #include <linux/platform_device.h>
 #include <linux/of.h>
 #include <soc/qcom/msm-clock-controller.h>
+#include <linux/reset-controller.h>
+
+static int dummy_reset_assert(struct reset_controller_dev *rcdev,
+				unsigned long id)
+{
+	return 0;
+}
+
+static int dummy_reset_deassert(struct reset_controller_dev *rcdev,
+				unsigned long id)
+{
+	return 0;
+}
+
+static int dummy_reset(struct reset_controller_dev *rcdev, unsigned long id)
+{
+	return 0;
+}
+
+struct reset_control_ops dummy_reset_ops = {
+	.reset = dummy_reset,
+	.assert = dummy_reset_assert,
+	.deassert = dummy_reset_deassert,
+};
+
+int dummy_reset_controller_register(struct platform_device *pdev)
+{
+	struct reset_controller_dev *prcdev;
+	int ret = 0;
+
+	prcdev = devm_kzalloc(&pdev->dev, sizeof(*prcdev), GFP_KERNEL);
+	if (!prcdev)
+		return -ENOMEM;
+
+	prcdev->of_node = pdev->dev.of_node;
+	prcdev->ops = &dummy_reset_ops;
+	prcdev->owner = pdev->dev.driver->owner;
+	prcdev->nr_resets = 17; // based on size of gcc_msm8996_resets
+
+	ret = reset_controller_register(prcdev);
+	if (ret)
+		dev_err(&pdev->dev, "%s: Failed to register with reset controller %d\n", __func__, ret);
+
+	return ret;
+}
 
 static int dummy_clk_reset(struct clk *clk, enum clk_reset_action action)
 {
@@ -98,10 +143,19 @@ static int msm_clock_dummy_probe(struct platform_device *pdev)
 	int ret;
 
 	ret = of_clk_add_provider(pdev->dev.of_node, of_dummy_get, NULL);
-	if (ret)
+	if (ret) {
+		dev_err(&pdev->dev, "%s: of_clk_add_provider error %d", __func__, ret);
 		return -ENOMEM;
+	}
+
+	ret = dummy_reset_controller_register(pdev);
+	if (ret) {
+		dev_err(&pdev->dev, "%s: dummy_reset_controller_register error %d", __func__, ret);
+		return -ENOMEM;
+	}
 
 	dev_info(&pdev->dev, "Registered DUMMY provider.\n");
+
 	return ret;
 }
 
diff --git a/drivers/platform/msm/gpio-usbdetect.c b/drivers/platform/msm/gpio-usbdetect.c
index 1628253..5d13689 100644
--- a/drivers/platform/msm/gpio-usbdetect.c
+++ b/drivers/platform/msm/gpio-usbdetect.c
@@ -46,6 +46,9 @@ static irqreturn_t gpio_usbdetect_vbus_irq(int irq, void *data)
 {
 	struct gpio_usbdetect *usb = data;
 
+	if (!usb->gpio)
+		return IRQ_HANDLED;
+
 	usb->vbus_state = gpio_get_value(usb->gpio);
 	if (usb->vbus_state) {
 		dev_dbg(&usb->pdev->dev, "setting vbus notification\n");
@@ -97,6 +100,7 @@ static int gpio_usbdetect_probe(struct platform_device *pdev)
 {
 	struct gpio_usbdetect *usb;
 	int rc;
+	bool skip_vbus_detect;
 
 	usb = devm_kzalloc(&pdev->dev, sizeof(*usb), GFP_KERNEL);
 	if (!usb)
@@ -136,68 +140,97 @@ static int gpio_usbdetect_probe(struct platform_device *pdev)
 		}
 	}
 
-	usb->gpio = of_get_named_gpio(pdev->dev.of_node,
-				"qcom,vbus-det-gpio", 0);
-	if (usb->gpio < 0) {
-		dev_err(&pdev->dev, "Failed to get gpio: %d\n", usb->gpio);
-		rc = usb->gpio;
-		goto error;
-	}
+	skip_vbus_detect = of_property_read_bool(pdev->dev.of_node,
+		"qcom,skip-vbus-detect");
+	if(!skip_vbus_detect) {
+		usb->gpio = of_get_named_gpio(pdev->dev.of_node,
+					"qcom,vbus-det-gpio", 0);
+		if (usb->gpio < 0) {
+			dev_err(&pdev->dev, "Failed to get gpio: %d\n", usb->gpio);
+			rc = usb->gpio;
+			goto error;
+		}
 
-	rc = gpio_request(usb->gpio, "vbus-det-gpio");
-	if (rc < 0) {
-		dev_err(&pdev->dev, "Failed to request gpio: %d\n", rc);
-		goto error;
-	}
+		rc = gpio_request(usb->gpio, "vbus-det-gpio");
+		if (rc < 0) {
+			dev_err(&pdev->dev, "Failed to request gpio: %d\n", rc);
+			goto error;
+		}
 
-	usb->vbus_det_irq = gpio_to_irq(usb->gpio);
-	if (usb->vbus_det_irq < 0) {
-		dev_err(&pdev->dev, "get vbus_det_irq failed\n");
-		rc = usb->vbus_det_irq;
-		goto error;
-	}
+		usb->vbus_det_irq = gpio_to_irq(usb->gpio);
+		if (usb->vbus_det_irq < 0) {
+			dev_err(&pdev->dev, "get vbus_det_irq failed\n");
+			rc = usb->vbus_det_irq;
+			goto error;
+		}
 
-	rc = devm_request_threaded_irq(&pdev->dev, usb->vbus_det_irq,
-				NULL, gpio_usbdetect_vbus_irq,
-			      IRQF_TRIGGER_RISING | IRQF_TRIGGER_FALLING |
-			      IRQF_ONESHOT, "vbus_det_irq", usb);
-	if (rc) {
-		dev_err(&pdev->dev, "request for vbus_det_irq failed: %d\n",
-			rc);
-		goto error;
-	}
+		rc = devm_request_threaded_irq(&pdev->dev, usb->vbus_det_irq,
+					NULL, gpio_usbdetect_vbus_irq,
+				      IRQF_TRIGGER_RISING | IRQF_TRIGGER_FALLING |
+				      IRQF_ONESHOT, "vbus_det_irq", usb);
+		if (rc) {
+			dev_err(&pdev->dev, "request for vbus_det_irq failed: %d\n",
+				rc);
+			goto error;
+		}
 
-	usb->id_det_irq = platform_get_irq_byname(pdev, "pmic_id_irq");
-	if (usb->id_det_irq < 0) {
-		dev_err(&pdev->dev, "get id_det_irq failed\n");
-		rc = usb->id_det_irq;
-		goto error;
-	}
+		usb->id_det_irq = platform_get_irq_byname(pdev, "pmic_id_irq");
+		if (usb->id_det_irq < 0) {
+			dev_err(&pdev->dev, "get id_det_irq failed\n");
+			rc = usb->id_det_irq;
+			goto error;
+		}
 
-	rc = devm_request_threaded_irq(&pdev->dev, usb->id_det_irq,
-				gpio_usbdetect_id_irq,
-				gpio_usbdetect_id_irq_thread,
-				IRQF_TRIGGER_RISING | IRQF_TRIGGER_FALLING |
-				IRQF_ONESHOT, "id_det_irq", usb);
-	if (rc) {
-		dev_err(&pdev->dev, "request for id_det_irq failed: %d\n", rc);
-		goto error;
+		rc = devm_request_threaded_irq(&pdev->dev, usb->id_det_irq,
+					gpio_usbdetect_id_irq,
+					gpio_usbdetect_id_irq_thread,
+					IRQF_TRIGGER_RISING | IRQF_TRIGGER_FALLING |
+					IRQF_ONESHOT, "id_det_irq", usb);
+		if (rc) {
+			dev_err(&pdev->dev, "request for id_det_irq failed: %d\n", rc);
+			goto error;
+		}
+
+		enable_irq_wake(usb->vbus_det_irq);
+		enable_irq_wake(usb->id_det_irq);
+	}
+	else
+	{
+		dev_info(&pdev->dev, "%s: skipping vbus_detect\n", __func__);
+		usb->vbus_det_irq = 0;
+		usb->id_det_irq = 0;
+		usb->gpio = 0;
 	}
 
-	enable_irq_wake(usb->vbus_det_irq);
-	enable_irq_wake(usb->id_det_irq);
 	dev_set_drvdata(&pdev->dev, usb);
 
-	if (usb->id_det_irq) {
-		gpio_usbdetect_id_irq(usb->id_det_irq, usb);
-		if (!usb->id_state) {
-			gpio_usbdetect_id_irq_thread(usb->id_det_irq, usb);
-			return 0;
+	if (!skip_vbus_detect) {
+		if (usb->id_det_irq) {
+			gpio_usbdetect_id_irq(usb->id_det_irq, usb);
+			if (!usb->id_state) {
+				gpio_usbdetect_id_irq_thread(usb->id_det_irq, usb);
+				return 0;
+			}
 		}
+
+		/* Read and report initial VBUS state */
+		gpio_usbdetect_vbus_irq(usb->vbus_det_irq, usb);
 	}
 
-	/* Read and report initial VBUS state */
-	gpio_usbdetect_vbus_irq(usb->vbus_det_irq, usb);
+	if(skip_vbus_detect && of_property_read_bool(pdev->dev.of_node,
+		"qcom,force-vbus-status-on"))
+	{
+		dev_info(&pdev->dev, "%s: force vbus status to ON\n", __func__);
+		extcon_set_cable_state_(usb->extcon_dev, EXTCON_USB_HOST, 0);
+		extcon_set_cable_state_(usb->extcon_dev, EXTCON_USB, 1);
+	}
+	else if(skip_vbus_detect && of_property_read_bool(pdev->dev.of_node,
+		"qcom,force-vbus-status-off"))
+	{
+		dev_info(&pdev->dev, "%s: force vbus status to OFF\n", __func__);
+		extcon_set_cable_state_(usb->extcon_dev, EXTCON_USB_HOST, 1);
+		extcon_set_cable_state_(usb->extcon_dev, EXTCON_USB, 0);
+	}
 
 	return 0;
 
diff --git a/drivers/usb/core/hcd.c b/drivers/usb/core/hcd.c
index 1fb9191..2badaa7 100644
--- a/drivers/usb/core/hcd.c
+++ b/drivers/usb/core/hcd.c
@@ -2696,6 +2696,10 @@ static int usb_hcd_request_irqs(struct usb_hcd *hcd,
 
 	if (hcd->driver->irq) {
 
+#ifdef CONFIG_MSM_GVM_QUIN
+		irqflags |= IRQF_ONESHOT;
+#endif
+
 		snprintf(hcd->irq_descr, sizeof(hcd->irq_descr), "%s:usb%d",
 				hcd->driver->description, hcd->self.busnum);
 		retval = request_irq(irqnum, &usb_hcd_irq, irqflags,
diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index 0744b14..a2c78be 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -1059,8 +1059,13 @@ static int dwc3_probe(struct platform_device *pdev)
 
 	/* will be enabled in dwc3_msm_resume() */
 	irq_set_status_flags(irq, IRQ_NOAUTOEN);
+#ifdef CONFIG_MSM_GVM_QUIN
+	ret = devm_request_irq(dev, irq, dwc3_interrupt, IRQF_SHARED|IRQF_ONESHOT, "dwc3",
+			dwc);
+#else
 	ret = devm_request_irq(dev, irq, dwc3_interrupt, IRQF_SHARED, "dwc3",
 			dwc);
+#endif
 	if (ret) {
 		dev_err(dwc->dev, "failed to request irq #%d --> %d\n",
 				irq, ret);
@@ -1311,7 +1316,9 @@ err0:
 	 * memory region the next time probe is called.
 	 */
 	res->start -= DWC3_GLOBALS_REGS_START;
-	destroy_workqueue(dwc->dwc_wq);
+	if (dwc->dwc_wq) {
+		destroy_workqueue(dwc->dwc_wq);
+	}
 
 	return ret;
 }
diff --git a/drivers/usb/dwc3/dwc3-msm.c b/drivers/usb/dwc3/dwc3-msm.c
index 0c992cf..20c7a28 100644
--- a/drivers/usb/dwc3/dwc3-msm.c
+++ b/drivers/usb/dwc3/dwc3-msm.c
@@ -1653,6 +1653,7 @@ static int dwc3_msm_link_clk_reset(struct dwc3_msm *mdwc, bool assert)
 		clk_disable_unprepare(mdwc->sleep_clk);
 		clk_disable_unprepare(mdwc->core_clk);
 		clk_disable_unprepare(mdwc->iface_clk);
+
 		ret = reset_control_assert(mdwc->core_reset);
 		if (ret)
 			dev_err(mdwc->dev, "dwc3 core_reset assert failed\n");
@@ -1994,6 +1995,8 @@ static void msm_dwc3_perf_vote_update(struct dwc3_msm *mdwc,
 
 static int dwc3_msm_suspend(struct dwc3_msm *mdwc)
 {
+
+#ifdef ENABLE_MSM_DWC3_USB_SUSPEND
 	int ret, i;
 	bool can_suspend_ssphy;
 	struct dwc3 *dwc = platform_get_drvdata(mdwc->dwc3);
@@ -2164,6 +2167,7 @@ static int dwc3_msm_suspend(struct dwc3_msm *mdwc)
 
 	dev_info(mdwc->dev, "DWC3 in low power mode\n");
 	mutex_unlock(&mdwc->suspend_resume_mutex);
+#endif
 	return 0;
 }
 
@@ -2200,6 +2204,7 @@ static int dwc3_msm_resume(struct dwc3_msm *mdwc)
 	if (mdwc->lpm_flags & MDWC3_POWER_COLLAPSE) {
 		dev_dbg(mdwc->dev, "%s: exit power collapse\n", __func__);
 		dwc3_msm_config_gdsc(mdwc, 1);
+
 		ret = reset_control_assert(mdwc->core_reset);
 		if (ret)
 			dev_err(mdwc->dev, "%s:core_reset assert failed\n",
@@ -2985,7 +2990,11 @@ static int dwc3_msm_probe(struct platform_device *pdev)
 		ret = devm_request_threaded_irq(&pdev->dev, mdwc->pwr_event_irq,
 					msm_dwc3_pwr_irq,
 					msm_dwc3_pwr_irq_thread,
+#ifdef CONFIG_MSM_GVM_QUIN
+					IRQF_TRIGGER_RISING | IRQF_EARLY_RESUME | IRQF_ONESHOT,
+#else
 					IRQF_TRIGGER_RISING | IRQF_EARLY_RESUME,
+#endif
 					"msm_dwc3", mdwc);
 		if (ret) {
 			dev_err(&pdev->dev, "irqreq pwr_event_irq failed: %d\n",
@@ -3361,6 +3370,7 @@ static int dwc3_msm_host_notifier(struct notifier_block *nb,
 
 			/* inform PMIC of max power so it can optimize boost */
 			pval.intval = max_power * 1000;
+
 			power_supply_set_property(mdwc->usb_psy,
 					POWER_SUPPLY_PROP_BOOST_CURRENT, &pval);
 		} else {
diff --git a/drivers/usb/host/xhci-plat.c b/drivers/usb/host/xhci-plat.c
index be72953..7ec0fe0 100644
--- a/drivers/usb/host/xhci-plat.c
+++ b/drivers/usb/host/xhci-plat.c
@@ -249,7 +249,11 @@ static int xhci_plat_probe(struct platform_device *pdev)
 			goto put_usb3_hcd;
 	}
 
+#ifdef CONFIG_MSM_GVM_QUIN
+	ret = usb_add_hcd(hcd, irq, IRQF_SHARED|IRQF_ONESHOT);
+#else
 	ret = usb_add_hcd(hcd, irq, IRQF_SHARED);
+#endif
 	if (ret)
 		goto disable_usb_phy;
 
@@ -258,7 +262,11 @@ static int xhci_plat_probe(struct platform_device *pdev)
 	if (HCC_MAX_PSA(xhci->hcc_params) >= 4)
 		xhci->shared_hcd->can_do_streams = 1;
 
+#ifdef CONFIG_MSM_GVM_QUIN
+	ret = usb_add_hcd(xhci->shared_hcd, irq, IRQF_SHARED|IRQF_ONESHOT);
+#else
 	ret = usb_add_hcd(xhci->shared_hcd, irq, IRQF_SHARED);
+#endif
 	if (ret)
 		goto dealloc_usb2_hcd;
 
diff --git a/drivers/usb/host/xhci.c b/drivers/usb/host/xhci.c
index 5e133de..51dafc6 100644
--- a/drivers/usb/host/xhci.c
+++ b/drivers/usb/host/xhci.c
@@ -426,8 +426,13 @@ static int xhci_try_enable_msi(struct usb_hcd *hcd)
 			 hcd->driver->description, hcd->self.busnum);
 
 	/* fall back to legacy interrupt*/
+#ifdef CONFIG_MSM_GVM_QUIN
+	ret = request_irq(pdev->irq, &usb_hcd_irq, IRQF_SHARED | IRQF_ONESHOT,
+			hcd->irq_descr, hcd);
+#else
 	ret = request_irq(pdev->irq, &usb_hcd_irq, IRQF_SHARED,
 			hcd->irq_descr, hcd);
+#endif
 	if (ret) {
 		xhci_err(xhci, "request interrupt %d failed\n",
 				pdev->irq);
diff --git a/drivers/usb/phy/phy-msm-ssusb-qmp.c b/drivers/usb/phy/phy-msm-ssusb-qmp.c
index 2bc3c6f..5ded50e 100644
--- a/drivers/usb/phy/phy-msm-ssusb-qmp.c
+++ b/drivers/usb/phy/phy-msm-ssusb-qmp.c
@@ -144,6 +144,11 @@ static void msm_ssusb_qmp_enable_autonomous(struct msm_ssphy_qmp *phy,
 	unsigned int autonomous_mode_offset =
 			phy->phy_reg[USB3_PHY_AUTONOMOUS_MODE_CTRL];
 
+	if(!phy->vls_clamp_reg)	{
+		dev_dbg(phy->phy.dev, "vls_clamp_reg not available\n");
+		return;
+	}
+
 	dev_dbg(phy->phy.dev, "enabling QMP autonomous mode with cable %s\n",
 			get_cable_status_str(phy));
 
@@ -641,16 +646,25 @@ static int msm_ssphy_qmp_probe(struct platform_device *pdev)
 		goto err;
 	}
 
-	res = platform_get_resource_byname(pdev, IORESOURCE_MEM,
-			"vls_clamp_reg");
-	if (!res) {
-		dev_err(dev, "failed getting vls_clamp_reg\n");
-		return -ENODEV;
+	if (!of_property_read_bool(dev->of_node, "qcom,disable-autonomous-mode")) {
+		dev_err(dev, "%s: autonomous mode enabled!\n", __func__);
+		res = platform_get_resource_byname(pdev, IORESOURCE_MEM,
+				"vls_clamp_reg");
+		if (!res) {
+			dev_err(dev, "failed getting vls_clamp_reg\n");
+			return -ENODEV;
+		}
+
+		phy->vls_clamp_reg = devm_ioremap_resource(dev, res);
+		if (IS_ERR(phy->vls_clamp_reg)) {
+			dev_err(dev, "couldn't find vls_clamp_reg address.\n");
+			return PTR_ERR(phy->vls_clamp_reg);
+		}
 	}
-	phy->vls_clamp_reg = devm_ioremap_resource(dev, res);
-	if (IS_ERR(phy->vls_clamp_reg)) {
-		dev_err(dev, "couldn't find vls_clamp_reg address.\n");
-		return PTR_ERR(phy->vls_clamp_reg);
+	else
+	{
+		dev_err(dev, "%s: autonomous mode disabled!\n", __func__);
+		phy->vls_clamp_reg = NULL;
 	}
 
 	res = platform_get_resource_byname(pdev, IORESOURCE_MEM,
-- 
1.9.1

